### PR TITLE
[#1035] Create ZcashError.unknown

### LINKED ISSUE #1035: Create ZcashError.unknown
Throwing functions throws `any Error` and our goal is to throw ZcashError in the whole SDK. The issue is that any app adopting SDK using throwing functions end up in 
```swift
do {
// ...
} catch error is ZcashError {
// all ok here
} catch {
// any Error here
}
```
Passing error in the apps would be complicated. We can improve this experience by presenting ZcashError.unknown and mapping any Error into it. 

That would result in:

```swift
do {
// ...
} catch {
guard error is ZcashError else { 
    throw error.mapToZcashError()
}
throw error 
}
```

### PR DESCRIPTION
**Pull request by LukasKorba**
Closes #1035 

- new ZcashError.unknown that wraps and handles any Error added

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

**Review by pacu — CHANGES_REQUESTED** [review]
My main objection to this is that it defeats the purpose of errors having locality and being pin pointed to there they occurred. 

Even a more local but yet generic error has some value of providing clients some code location.

Being reported an unknown error would be undesirable to us, because:
1- our app throws unknown errors.
2- we don't get any information beforehand that the user can capture to fix it.

**Comment by Chlup** [comment]
> My main objection to this is that it defeats the purpose of errors having locality and being pin pointed to there they occurred.
> 
> Even a more local but yet generic error has some value of providing clients some code location.
> 
> Being reported an unknown error would be undesirable to us, because: 1- our app throws unknown errors. 2- we don't get any information beforehand that the user can capture to fix it.

I agree. Problem is that our SDK emits and throws `Error` type. Not `ZcashError`. It may be possible to change error types in streams. But it's impossible to specify error type when throwing error from method. So we need some easy way how to convert `Error` type to `ZcashError`. Because in fact all the errors generated by the SDK are `ZcashError`. Without `unknown` error there will be some Optional handling when doing `Error` -> `ZcashError` conversion. `unknown` is there to not need to handle Optionals when doing that conversion. And it should never be used directly. Like no function should explicitly throw `ZcashError.unknown`. It will be used only in case that we made a mistake and somewhere is thrown something else than `ZcashError`. And it will be used only in case that client uses `toZcashError()` function. 

I checked the PR and `toZcashError()` is wrongly implemented due to misunderstanding. Implementation should looks like this:

**Comment by LukasKorba** [comment]
Main reason we added this is to use TCA actions
```swift
case someFailure(ZcashError)
```

so we need to ensure ZcashError is possible to pass from a throwing function. If it's not a ZcashError, we don't have a tool to wrap it from the app side.

**Comment by pacu** [comment]
I guess we don't have much wiggle room here and will have to live with it.

File: Sources/ZcashLightClientKit/Extensions/Error+ZcashError.swift [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+//  Error+ZcashError.swift
+//  
+//
+//  Created by Lukáš Korba on 09.05.2023.
+//
+
+import Foundation
+
+public extension Error {
+    func toZcashError() -> ZcashError {
```
Comment (pacu) [inline Sources/ZcashLightClientKit/Extensions/Error+ZcashError.swift:11]: Also explain when this should be used and why

**Pull request merged**